### ISSUE TITLE
css/filter-effects /backdrop-filter-edge-behavior.html seems to test incorrect behaviour

### ISSUE THREAD

**Issue by mattwoodrow** [web-platform-tests/wpt#42071]
It looks like this test is setting a backdrop-filter:blur within a scroller, and is expecting the border of the scroller to not be included in the blurred backdrop.

`overflow:scroll` isn't a backdrop root, so isn't having the red border pixels included the expected behaviour?

@mfreed7

**Comment by mfreed7** [comment]
The border is painted on top/outside of the scrolled content, which contains the backdrop filter. So the red border should never be **behind** the backdrop filter, and should therefore not filter any red content.

**Comment by mattwoodrow** [comment]
Ah, I was expecting the border to be painted before the backdrop.

Looking at https://www.w3.org/TR/CSS21/zindex.html I would expect the border of ‘scroller’ to be painted during step 4.3 of the root element stacking context, whereas the ‘backdrop’ element’s stacking context would be painted during step 8.

**Comment by mattwoodrow** [comment]
Could you take another look at this please @mfreed7?

I can see in the Chrome dev tools layer inspector, that the red border is indeed in the document layer, behind the backdrop layer.

My best guess is that the 'Backdrop Root Image' is being clipped by the CSS `overflow` clip, such that the image to be blurred doesn't have any red pixels in it.

https://drafts.fxtf.org/filter-effects-2/#backdrop-filter-operation says:

> If the filter functions list includes a [blur()](https://drafts.fxtf.org/filter-effects-1/#blurEquivalent) filter, the filter will be applied with edgeMode="mirror", with the edge defined by the clipped, transformed border box of the element. See [§ 3 Backdrop Root](https://drafts.fxtf.org/filter-effects-2/#BackdropRoot).

 which sounds like it might? mean this, but the referred to backdrop root section is fairly explicit with:

> Transform the border box of element E to 2D screen-space, and clip the final painted output to this border quad.

As the only mention of clipping.

The border-box of "backdrop" transformed into screen space is definitely a quad that includes red pixels, and thus I think they should be blurring into the final result.

**Comment by mfreed7** [comment]
@progers WDYT?

**Comment by progers** [comment]
I agree that the spec says that the border should be included. It looks like matching Safari/WebKit was one of the original reasons for this change in https://github.com/w3c/fxtf-drafts/pull/342.

**Comment by progers** [comment]
@ogerchikov , just FYI (this will likely require a small change in chromium)

**Comment by gw3583** [comment]
@mattwoodrow  Should WPT be updated to have Chrome/Edge marked as failing for this test then, with Safari as the passing test? Firefox looks like it matches Safari behaviour fairly closely, though will likely need an extra CSS style to hide the scrollbar (current test has `-webkit-scrollbar`) I think.

**Comment by mattwoodrow** [comment]
Yeah, I think we can adjust the reference file to show the blurred red.

**Comment by gw3583** [comment]
> Yeah, I think we can adjust the reference file to show the blurred red.

@mattwoodrow  It's not quite clear to me what we should be testing for in this case now (i.e. what would be a reasonable reference file that still tests something useful). Would you mind updating the reference file for it?

**Comment by Maya-Fey** [comment]
> If the filter functions list includes a [blur()](https://drafts.fxtf.org/filter-effects-1/#blurEquivalent) filter, the filter will be applied with edgeMode="mirror", with the edge defined by the clipped, transformed border box of the element. See [§ 3 Backdrop Root](https://drafts.fxtf.org/filter-effects-2/#BackdropRoot).

As it is currently implemented in chromium, the edge is defined by the `visible_rect` of the draw quad of the element with the backdrop filter (ie, 'element B' in the spec), For an element that's fully visible, this should be the border box if I understand correctly. I thought at first this was the correct behavior, ie "the element" = "element B", but, I think the intended reading is that "the element" that defines the edge is the element that defines the backdrop root. If this is the case, and I think that it is, I think the sentence should be rewritten to make it explicit that the element is the backdrop root, and the backdrop root isn't just being indirectly referenced in service of explaining the concept.

Given that, the WebKit behavior with this test is entirely correct, and chrome should be changed to match to fit the spec. However, actually implementing this leads to the question of what to do about [crbug.com/40739693](https://issues.chromium.org/issues/40739693). The reason we crop to the `visible_rect` is to mitigate this issue. You can see that on firefox, this issue still repros, see https://codepen.io/Bes7weB/pen/MWjQMMP (I do not have a machine with which to test current safari, perhaps webkit has its own fix for this).

My [(very) rough revert](https://chromium-review.googlesource.com/c/chromium/src/+/6818838) of our behavior reintroduces the issue, although it does fix the test and match WebKit behavior, mostly by coincidence (using the border rect as the edge of the blur filter is not implemented, but the resulting rect ends up being big enough to include the edges the test is testing for). 

I don't think reintroducing this bug is a good idea, it's obviously not the correct behavior, even by spec. I believe what we *should* be doing is taking only into account the painted elements between the root and the filtered element, but it seems in practicea bunch of other things are included in the filtered image as well. FF seems to have the same issue as mentioned. I'm still learning the viz internals so I'm not quite sure as to the cause. So I suppose my question is, is this an easy fix, or would it be worthwhile to codify the visible_rect behavior, maybe using the visible rect of the root rather than the filtered element, which I think would pass both the WPT test and prevent this bug, in order to ease implementation burden?

@progers WDYT?
